src/SonarScanner.MSBuild.PreProcessor/JreCaching/UnpackerFactory.cs

```diff
@@ -0,0 +1,37 @@
+﻿/*
+ * SonarScanner for .NET
+ * Copyright (C) 2016-2024 SonarSource SA
+ * mailto: info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.IO;
+using SonarScanner.MSBuild.Common;
+
+namespace SonarScanner.MSBuild.PreProcessor.JreCaching;
+
+public class UnpackerFactory : IUnpackerFactory
+{
+    public static UnpackerFactory Instance { get; } = new UnpackerFactory();
+
+    public IUnpacker CreateForArchive(IDirectoryWrapper directoryWrapper, IFileWrapper fileWrapper, IOperatingSystemProvider operatingSystemProvider, string archive) =>
+        Path.GetExtension(archive).ToUpperInvariant() switch
+        {
+            ".ZIP" => new ZipUnpacker(),
+            ".GZ" => new TarGzUnpacker(directoryWrapper, fileWrapper, operatingSystemProvider),
```

Comment:
Check also for .tar with the help of `Path.GetFileNameWithoutExtension`:
```suggestion
            ".GZ" when string.Equals(Path.GetExtension(Path.GetFileNameWithoutExtension(archive)), ".tar", StringComparison.OrdinalIgnoreCase) => new TarGzUnpacker(directoryWrapper, fileWrapper, operatingSystemProvider),
```